From 3cf88028e64ce275118f8071fde2baa5a2a0ccf9 Mon Sep 17 00:00:00 2001
From: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>
Date: Fri, 21 Jul 2023 13:23:14 -0400
Subject: [PATCH 1/2] add Python 3.12 support to ansible-test (#80834)

* update docker containers versions to use newer ansible-test ref in the pre-built venvs

* Allow invoking ansible-test with Python 3.12

* Add python3.12 to the INTERPRETER_PYTHON_FALLBACK

* changelog

* add Python 3.12 as a non-default Python version for the test containers

* Update mypy ignores for Python 3.12

* Add Python 3.12 to CI matrix for unit tests, generic tests, and galaxy

* Update unit test for using the Python 2 collection loader path with Python 3.

Skip the existing test on Python 3.12, since find_module is removed.

Suppress the pre-existing deprecation warnings using the Python 2
codepath with Python 3.

Add a test for Python >= 3.12, which doesn't call find_module.

* Ignore sanity test errors on systems without libselinux present.
---
 lib/ansible/config/base.yml                   |  1 +
 .../ansible_test/_data/completion/docker.txt  |  6 ++---
 .../_internal/python_requirements.py          |  4 ++--
 .../_util/target/common/constants.py          |  1 +
 .../_util/target/setup/bootstrap.sh           |  2 +-
 test/sanity/ignore.txt                        |  6 +++--
 .../test_collection_loader.py                 | 22 +++++++++++++++++--
 7 files changed, 32 insertions(+), 10 deletions(-)

diff --git a/lib/ansible/config/base.yml b/lib/ansible/config/base.yml
index 052a8f0834..64e1099059 100644
--- a/lib/ansible/config/base.yml
+++ b/lib/ansible/config/base.yml
@@ -1548,6 +1548,7 @@ _INTERPRETER_PYTHON_DISTRO_MAP:
 INTERPRETER_PYTHON_FALLBACK:
   name: Ordered list of Python interpreters to check for in discovery
   default:
+  - python3.12
   - python3.11
   - python3.10
   - python3.9
diff --git a/test/lib/ansible_test/_data/completion/docker.txt b/test/lib/ansible_test/_data/completion/docker.txt
index ee3f082c1f..f73fdb408e 100644
--- a/test/lib/ansible_test/_data/completion/docker.txt
+++ b/test/lib/ansible_test/_data/completion/docker.txt
@@ -1,6 +1,6 @@
-base image=quay.io/ansible/base-test-container:4.1.0 python=3.11,2.7,3.5,3.6,3.7,3.8,3.9,3.10
-default image=quay.io/ansible/default-test-container:7.14.0 python=3.11,2.7,3.5,3.6,3.7,3.8,3.9,3.10 context=collection
-default image=quay.io/ansible/ansible-core-test-container:7.14.0 python=3.11,2.7,3.5,3.6,3.7,3.8,3.9,3.10 context=ansible-core
+base image=quay.io/ansible/base-test-container:5.0.0 python=3.11,2.7,3.6,3.7,3.8,3.9,3.10,3.12
+default image=quay.io/ansible/default-test-container:8.3.0 python=3.11,2.7,3.6,3.7,3.8,3.9,3.10,3.12 context=collection
+default image=quay.io/ansible/ansible-core-test-container:8.3.0 python=3.11,2.7,3.6,3.7,3.8,3.9,3.10,3.12 context=ansible-core
 alpine3 image=quay.io/ansible/alpine3-test-container:5.0.0 python=3.10 cgroup=none audit=none
 centos7 image=quay.io/ansible/centos7-test-container:5.0.0 python=2.7 cgroup=v1-only
 fedora37 image=quay.io/ansible/fedora37-test-container:5.0.0 python=3.11
diff --git a/test/lib/ansible_test/_internal/python_requirements.py b/test/lib/ansible_test/_internal/python_requirements.py
index 506b802c52..adf9445161 100644
--- a/test/lib/ansible_test/_internal/python_requirements.py
+++ b/test/lib/ansible_test/_internal/python_requirements.py
@@ -441,8 +441,8 @@ def get_venv_packages(python: PythonConfig) -> dict[str, str]:
     #       See: https://github.com/ansible/base-test-container/blob/main/files/installer.py
 
     default_packages = dict(
-        pip='21.3.1',
-        setuptools='60.8.2',
+        pip='23.1.2',
+        setuptools='67.7.2',
         wheel='0.37.1',
     )
 
diff --git a/test/lib/ansible_test/_util/target/common/constants.py b/test/lib/ansible_test/_util/target/common/constants.py
index 9bddfaf439..f3c3857ef9 100644
--- a/test/lib/ansible_test/_util/target/common/constants.py
+++ b/test/lib/ansible_test/_util/target/common/constants.py
@@ -17,4 +17,5 @@ CONTROLLER_PYTHON_VERSIONS = (
     '3.9',
     '3.10',
     '3.11',
+    '3.12',
 )
diff --git a/test/lib/ansible_test/_util/target/setup/bootstrap.sh b/test/lib/ansible_test/_util/target/setup/bootstrap.sh
index ea17dad387..367dcfcb4c 100644
--- a/test/lib/ansible_test/_util/target/setup/bootstrap.sh
+++ b/test/lib/ansible_test/_util/target/setup/bootstrap.sh
@@ -53,7 +53,7 @@ install_pip() {
                 pip_bootstrap_url="https://ci-files.testing.ansible.com/ansible-test/get-pip-20.3.4.py"
                 ;;
             *)
-                pip_bootstrap_url="https://ci-files.testing.ansible.com/ansible-test/get-pip-21.3.1.py"
+                pip_bootstrap_url="https://ci-files.testing.ansible.com/ansible-test/get-pip-23.1.2.py"
                 ;;
         esac
 
diff --git a/test/sanity/ignore.txt b/test/sanity/ignore.txt
index f62fcbfc09..4d3ffc04b9 100644
--- a/test/sanity/ignore.txt
+++ b/test/sanity/ignore.txt
@@ -9,12 +9,13 @@ lib/ansible/executor/task_queue_manager.py pylint:disallowed-name
 lib/ansible/galaxy/collection/__init__.py mypy-3.9:attr-defined  # inline ignore has no effect
 lib/ansible/galaxy/collection/__init__.py mypy-3.10:attr-defined  # inline ignore has no effect
 lib/ansible/galaxy/collection/__init__.py mypy-3.11:attr-defined  # inline ignore has no effect
-lib/ansible/galaxy/collection/gpg.py mypy-3.9:arg-type
+lib/ansible/galaxy/collection/__init__.py mypy-3.12:attr-defined  # inline ignore has no effect
 lib/ansible/galaxy/collection/gpg.py mypy-3.10:arg-type
 lib/ansible/galaxy/collection/gpg.py mypy-3.11:arg-type
-lib/ansible/parsing/yaml/constructor.py mypy-3.9:type-var  # too many occurrences to ignore inline
+lib/ansible/galaxy/collection/gpg.py mypy-3.12:arg-type
 lib/ansible/parsing/yaml/constructor.py mypy-3.10:type-var  # too many occurrences to ignore inline
 lib/ansible/parsing/yaml/constructor.py mypy-3.11:type-var  # too many occurrences to ignore inline
+lib/ansible/parsing/yaml/constructor.py mypy-3.12:type-var  # too many occurrences to ignore inline
 lib/ansible/keyword_desc.yml no-unwanted-files
 lib/ansible/modules/apt.py validate-modules:parameter-invalid
 lib/ansible/modules/apt_repository.py validate-modules:parameter-invalid
@@ -77,6 +78,7 @@ lib/ansible/module_utils/compat/selinux.py import-3.8!skip # pass/fail depends o
 lib/ansible/module_utils/compat/selinux.py import-3.9!skip # pass/fail depends on presence of libselinux.so
 lib/ansible/module_utils/compat/selinux.py import-3.10!skip # pass/fail depends on presence of libselinux.so
 lib/ansible/module_utils/compat/selinux.py import-3.11!skip # pass/fail depends on presence of libselinux.so
+lib/ansible/module_utils/compat/selinux.py import-3.12!skip # pass/fail depends on presence of libselinux.so
 lib/ansible/module_utils/distro/_distro.py future-import-boilerplate # ignore bundled
 lib/ansible/module_utils/distro/_distro.py metaclass-boilerplate # ignore bundled
 lib/ansible/module_utils/distro/_distro.py no-assert
diff --git a/test/units/utils/collection_loader/test_collection_loader.py b/test/units/utils/collection_loader/test_collection_loader.py
index 9be59307a9..feaaf97a9c 100644
--- a/test/units/utils/collection_loader/test_collection_loader.py
+++ b/test/units/utils/collection_loader/test_collection_loader.py
@@ -29,8 +29,16 @@ def teardown(*args, **kwargs):
 # BEGIN STANDALONE TESTS - these exercise behaviors of the individual components without the import machinery
 
 
-@pytest.mark.skipif(not PY3, reason='Testing Python 2 codepath (find_module) on Python 3')
-def test_find_module_py3():
+@pytest.mark.filterwarnings(
+    'ignore:'
+    r'find_module\(\) is deprecated and slated for removal in Python 3\.12; use find_spec\(\) instead'
+    ':DeprecationWarning',
+    'ignore:'
+    r'FileFinder\.find_loader\(\) is deprecated and slated for removal in Python 3\.12; use find_spec\(\) instead'
+    ':DeprecationWarning',
+)
+@pytest.mark.skipif(not PY3 or sys.version_info >= (3, 12), reason='Testing Python 2 codepath (find_module) on Python 3, <= 3.11')
+def test_find_module_py3_lt_312():
     dir_to_a_file = os.path.dirname(ping_module.__file__)
     path_hook_finder = _AnsiblePathHookFinder(_AnsibleCollectionFinder(), dir_to_a_file)
 
@@ -40,6 +48,16 @@ def test_find_module_py3():
     assert path_hook_finder.find_module('missing') is None
 
 
+@pytest.mark.skipif(sys.version_info < (3, 12), reason='Testing Python 2 codepath (find_module) on Python >= 3.12')
+def test_find_module_py3_gt_311():
+    dir_to_a_file = os.path.dirname(ping_module.__file__)
+    path_hook_finder = _AnsiblePathHookFinder(_AnsibleCollectionFinder(), dir_to_a_file)
+
+    # setuptools may fall back to find_module on Python 3 if find_spec returns None
+    # see https://github.com/pypa/setuptools/pull/2918
+    assert path_hook_finder.find_spec('missing') is None
+
+
 def test_finder_setup():
     # ensure scalar path is listified
     f = _AnsibleCollectionFinder(paths='/bogus/bogus')
-- 
2.42.0

